core/src/com/unciv/logic/map/MapUnit.kt

```diff
@@ -528,22 +569,49 @@ class MapUnit {
             return canUpgrade
         }
         return false
     }
 
-    fun getCostOfUpgrade(): Int {
-        val unitToUpgradeTo = getUnitToUpgradeTo()
-        var goldCostOfUpgrade = (unitToUpgradeTo.cost - baseUnit().cost) * 2f + 10f
-        for (unique in civInfo.getMatchingUniques(UniqueType.UnitUpgradeCost, StateForConditionals(civInfo, unit=this)))
-            goldCostOfUpgrade *= unique.params[0].toPercent()
-
-        if (goldCostOfUpgrade < 0) return 0 // For instance, Landsknecht costs less than Spearman, so upgrading would cost negative gold
-        return goldCostOfUpgrade.toInt()
+    /** Determine gold cost of a Unit Upgrade, potentially over several steps.
+     *  @param unitToUpgradeTo the final BaseUnit. Must be reachable via normal upgrades or else
+     *         the function will return the cost to upgrade to the last possible and researched normal upgrade.
+     *  @return Gold cost in increments of 5, never negative. Will return 0 for invalid inputs (unit can't upgrade or is is already a [unitToUpgradeTo])
+     */
+    // Only one use from getUpgradeAction at the moment, so AI-specific rules omitted
+    //todo Does the AI never buy upgrades???
+    fun getCostOfUpgrade(unitToUpgradeTo: BaseUnit): Int {
+        // Reference: CvUnit.cpp Line 9357 (https://sourceforge.net/p/faerun/code-0/670/tree/Faerun%20DLL/CvGameCoreDLL_Expansion2/CvUnit.cpp)
+        // Source rounds to int every step, we don't
+        // From the source, this should apply _Production_ modifiers (Temple of Artemis? GameSpeed! StartEra!), at the moment it doesn't
+        var goldCostOfUpgrade = 0
+
+        val ruleset = civInfo.gameInfo.ruleSet
+        val constants = ruleset.modOptions.constants.unitUpgradeCost
+        // apply modifiers: Wonders (Pentagon), Policies (Professional Army). Cached outside loop despite
+        // the UniqueType being allowed on a BaseUnit - we don't have a MapUnit in the loop.
+        // Actually instantiating every intermediate to support such mods: todo
+        var civModifier = 1f
+        val stateForConditionals = StateForConditionals(civInfo, unit = this)
+        for (unique in civInfo.getMatchingUniques(UniqueType.UnitUpgradeCost, stateForConditionals))
+            civModifier *= unique.params[0].toPercent()
+
+        followUpgradePath { oldUnit, newUnit ->
+            // do clamping and rounding here so upgrading stepwise costs the same as upgrading far down the chain
+            var stepCost = constants.base
+            stepCost += (constants.perProduction * (newUnit.cost - oldUnit.cost)).coerceAtLeast(0f)
+            val era = ruleset.eras[ruleset.technologies[newUnit.requiredTech]?.era()]
+            if (era != null)
+                stepCost *= (1f + era.eraNumber * constants.eraMultiplier)
+            stepCost = (stepCost * civModifier).pow(constants.exponent)
+            goldCostOfUpgrade += (stepCost / constants.roundTo).toInt() * constants.roundTo
```

Comment:
... and project-wide unifying of, say, formatting a Float to 1 decimal place, or.. or.. or.. Who will rise to The Challenge? Global ExtensionFunctions reorg? C++ did `/=` `*=`! This here could probably be also written using stdlib stuff, but - honestly I didn't really check, as the first fit that came to mind rounds to nearest instead of down.

Reply:
> major brain power

You're `<b>`bold`</b>`, inviting nasty sarcasm like that :zany_face: 

But I invested a minute or two, with the goal (Float->Int + multiple of + round _down_) `/ *` is the simplest using stdlibs only.